### PR TITLE
updates notebooks for multistage with subgraphs

### PR DESCRIPTION
**Pull request by jperez999**
This PR updates the multi stage recsys notebooks to use subgraph. There are quite a few changes that occur in there. Some major changes include a shift in flow of what information is stored in the feature store. Now the feature store holds raw item and raw user information. These values have not been pre processed. So when they are retrieved in the systems graph they must go through a preprocessing step. In the first notebook, we add usage of the Subgraph operator, and we create to subgraphs one for item and one for user. We also create another subgraph for the item categorification. This is so that we can categorify the item_features separately when they are used to retrieve item embeddings. In the second notebook, the use of subgraphs forces the ensemble to introduce NVT workflows to handle the preprocessing of the data after it is retrieved for both users and items. Here is where we introduce the usage of subworkflows, which are based on subgraph.  

This PR depends on the following PRs: 
https://github.com/NVIDIA-Merlin/core/pull/349
https://github.com/NVIDIA-Merlin/systems/pull/372 
https://github.com/NVIDIA-Merlin/core/pull/350
https://github.com/NVIDIA-Merlin/core/pull/353
https://github.com/NVIDIA-Merlin/systems/pull/378

### CODE REVIEW

File: examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb [hunk 1]
```diff
@@ -7,7 +7,7 @@
    "metadata": {},
```
Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <p>why did we change feast &lt; 0.20 again? I thought we can now use <code>0.31</code> version? Feast 0.19 was problematic with dask_cudf. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <p>may be we can add some text right above this cell to explain what we are doing here and what we are registering to feast repo. I can do that.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <p>we need to explain this cell what's going on here, since <code>cat_wkflow = nvt_wkflow.get_subworkflow("items_cat")</code> is a new syntax.. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='69'/>

File: examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb [hunk 2]
```diff
@@ -7,7 +7,7 @@
    "metadata": {},
```
Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: <pre class="ql-syntax" spellcheck="false">Line #5.        view="user_features",
</pre><p>guess <code>view</code> should be <code>user_attributes</code> ?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: <pre class="ql-syntax" spellcheck="false">Line #3.        view="item_features",
</pre><p>similarly, here <code>view</code> should be <code>item_attributes</code>?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

**Comment by rnyak** [comment]
@jperez999 I provided some initial comments, but did not test the notebooks at my end yet, I will test latest on Monday.

I am concerned about `feast` version. I had issues when I installed `feast<0.20` since it was messing up dask_cudf and I could not run anything after that point.  Arent we ok with feast `0.31` why we need older feast version again?

File: examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb (hunk 1 above)
Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <pre class="ql-syntax" spellcheck="false">Line #14.    os.environ["TF_GPU_ALLOCATOR"] = "cuda_malloc_async"
</pre><p>guess we need to move this line right after <code>import os</code> to make it effective. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <pre class="ql-syntax" spellcheck="false">Line #5.    item_embs_df["item_id"] = item_features["item_id"]
</pre><p>I am skeptical that this step is correct.  This is not doing correct mapping.  if we are trying to replace encoded items with their original raw item_ids, we should use <code>unique.item_id.parquet</code>. so we can do something like </p><p> </p><pre class="ql-syntax" spellcheck="false">raw_item_ids = pd.read_parquet('./categories/unique.item_id.parquet')
item_embs_df["item_id"] = raw_item_ids["item_id"]  
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='70'/>

Comment (rnyak) [reply examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <p>In addition, are we sure we want to save item_embeddings for raw item_ids? Arent we supposed to do KNN search after retrieval stage based on the encoded items? If yes, we should then get rid of the steps above.</p>

File: examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb (hunk 2 above)
Comment (rnyak) [reply examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: @jperez999 and @karlhigley changing view to `user_attributes` give error since it was saved to feature store as `user_features`. so we should leave it as it is.

File: examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb (hunk 1 above)
Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <pre class="ql-syntax" spellcheck="false">Line #1.    user_id_raw = ["user_id"] &gt;&gt; Rename(postfix='_raw') &gt;&gt; LambdaOp(lambda col: col.astype("int32")) &gt;&gt; TagAsUserFeatures()
</pre><p>I think we no longer need <code>user_id_raw</code> and <code>item_id_raw</code> since we do not register them in the feature store.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

File: examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb (hunk 2 above)
Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: <pre class="ql-syntax" spellcheck="false">Line #2.        store=feature_store,
</pre><p>I am not sure how this step works without registering any encoded ids in the feature store. I assume <code>retrieval["candidate_ids"]&nbsp;</code> returns encoded item ids? then, how feature store can map encoded ids to the raw ids if it does not have such mapping info stored?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

File: examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb (hunk 1 above)
Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <pre class="ql-syntax" spellcheck="false">Line #6.    schema = train.schema.without(['user_id_raw', 'item_id_raw'])
</pre><p>if we remove <code>user_id_raw</code> and <code>item_id_raw</code> above from NVT pipeline, we can remove <code>.without</code> here as well.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

File: examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb (hunk 2 above)
Comment (jperez999) [reply examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: yeah you cant change the name of the view, that comes from the feast configs at the bottom of notebook 1.

Comment (jperez999) [reply examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: Same as above for this change. You cannot make this change unless we change the configs.

File: examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb (hunk 1 above)
Comment (jperez999) [reply examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: You need to save the embeddings with raw_item_ids because those embeddings should correspond to raw data. This way when you do the lookup in the second notebook you are getting the actual items. Those items are then processed through the item subworkflow to get the features for the model to rank the items. Suppose you wanted to use that similarity search with a different feature engineering workflow, if you had placed the data with the item ids already categorified that would not work. For the item_id to raw_item_ids you are correct that it is not correctly merged. However, I pose the question why are we sorting IDs by default? I see that we use it in the code for categorify to handle making na/nulls the first items in the dataframe... but as a user I think I expect to see the categories given in a first come first serve order. Unless I elect to use freq_threshold or kwarg that changes the strategy. Anyway I will fix for now.

Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <p>Let's rephrase this as</p><p> </p><p>"We start with the offline candidate retrieval stage. We are going to train a Two-Tower model for item retrieval using only positive interactions, that's why we apply Filter operator below. To learn ..."</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

**Comment by rnyak** [comment]
@jperez999 we need to update the unit test as well..

File: examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb (hunk 1 above)
Comment (rnyak) [reply examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: @jperez999 I wanted to say if `os.environ["TF_GPU_ALLOCATOR"] = "cuda_malloc_async"` comes after `import tensorflow` it might work.

File: examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb (hunk 2 above)
Comment (karlhigley) [reply examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: You can change it if it’s changed in both places though, right?

Comment (karlhigley) [reply examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: I think we probably should change both for the sake of clarity

File: examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb (hunk 1 above)
Comment (karlhigley) [reply examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: I think it has to happen in with the `os` call first and `import tensorflow` second so that TF picks up the config from the environment when it loads

File: examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb (hunk 2 above)
Comment (rnyak) [reply examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: @karlhigley agreed :)

File: examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb (hunk 1 above)
Comment (rnyak) [reply examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <p>I say let's move <code>os.environ["TF_GPU_ALLOCATOR"] = "cuda_malloc_async"</code> right after <code>import os</code> based on my experience :)   </p>

Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <pre class="ql-syntax" spellcheck="false">Line #3.    )
</pre><p>so we can save this as&nbsp;<code>user_attributes</code>&nbsp;if you want like below:</p><p> </p><p><code>user_features.to_parquet(</code></p><p><code>&nbsp;&nbsp;os.path.join(feature_repo_path, "data", "user_attributes.parquet")</code></p><p><code>)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <p>so we can save this as <code>item_attributes</code> if you want like below:</p><p><code></code><code></p><p>item_features.to_parquet(</p><p>&nbsp;&nbsp;os.path.join(feature_repo_path, "data", "item_attributes.parquet")</p><p>)</p><p></code><code></code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

File: examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb (hunk 2 above)
Comment (karlhigley) [reply examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: I took a stab at adding string id columns to the dataset generation this morning, and quickly realized there are some bigger issues to sort out re: synthetic data and how our libraries relate to it. @oliverholworthy is going to take a look and try to figure out where we should go with that, but for now I think we're going to have to continue with integer IDs. I wonder if we could make them more visually distinct from the Categorified IDs if we added a large offset to the raw IDs (e.g. `1_000_000`)?

Comment (rnyak) [reply examples/Building-and-deploying-multi-stage-RecSys/02-Deploying-multi-stage-RecSys-with-Merlin-Systems.ipynb:7]: > I took a stab at adding string id columns to the dataset generation this morning, and quickly realized there are some bigger issues to sort out re: synthetic data and how our libraries relate to it. @oliverholworthy is going to take a look and try to figure out where we should go with that, but for now I think we're going to have to continue with integer IDs. I wonder if we could make them more visually distinct from the Categorified IDs if we added a large offset to the raw IDs (e.g. `1_000_000`)?

@karlhigley  if this is a `nice to have` things we dont have to prioritize it now.  you call :)

**Review by karlhigley — APPROVED** [review]
Looks good from the perspective of code changes. @rnyak, let’s update the text in a second PR

File: tests/unit/examples/test_building_deploying_multi_stage_RecSys.py [hunk 3]
```diff
@@ -74,7 +74,7 @@ def test_func(tmpdir):
             df_lib = get_lib()
             train = df_lib.read_parquet(
                 os.path.join("{tmpdir / "data"}/processed_nvt/", "train", "part_0.parquet"),
```
Comment (rnyak) [inline tests/unit/examples/test_building_deploying_multi_stage_RecSys.py:76]: is that sending a request of transformed dataset instead of raw user_id?

File: examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb (hunk 1 above)
Comment (rnyak) [inline examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: <pre class="ql-syntax" spellcheck="false">Line #4.    workflow =  nvt.Workflow(["item_id"] + (['item_id', 'item_brand', 'item_category', 'item_shop'] &gt;&gt; TransformWorkflow(nvt_wkflow.get_subworkflow("item")) &gt;&gt; PredictTensorflow(model_tt.first.item_block())))
</pre><p>This step is not really intuitive from user perspective. we should explain why we have two times <code>item_id</code>?  why we need to do <code>fit_transform</code> not only <code>transform</code>? why are we fitting again?   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/Merlin/pull/1022/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='69'/>

Comment (karlhigley) [reply examples/Building-and-deploying-multi-stage-RecSys/01-Building-Recommender-Systems-with-Merlin.ipynb:7]: This line and comment have us way down a rabbit-hole fixing underlying issues that make this awkward 😅
